spinoso-string/src/lib.rs

```diff
@@ -2373,183 +2007,7 @@ impl String {
-                            // Push `endth` more positive because this match has
-                            // the effect of shrinking `slice`.
-                            let endth = substr.len() - slice.len() + size;
-                            return Some(&substr[..endth]);
-                        }
-                        // Otherwise, keep track of the character we observed and
-                        // advance the slice to continue decoding.
-                        (Some(_), size) => {
-                            slice = &slice[size..];
-                            remaining -= 1;
-                        }
-
-                        // The next two arms handle the case where we have
-                        // encountered an invalid UTF-8 byte sequence.
-                        //
-                        // In this case, `decode_utf8` will return slices whose
-                        // length is `1..=3`. The length of this slice is the
-                        // number of "characters" we can advance the loop by.
-                        //
-                        // If the invalid UTF-8 sequence contains more bytes
-                        // than we have remaining to get to the `end`th char,
-                        // then we can break the loop directly.
-                        (None, size) if remaining <= size => {
-                            // For an explanation of this arithmetic:
-                            // If we're trying to slice:
-                            //
-                            // ```
-                            // s = "a\xF0\x9F\x87"
-                            // s[0, 2]
-                            // ```
-                            //
-                            // By the time we get to this branch in this loop:
-                            //
-                            // ```
-                            // substr = "a\xF0\x9F\x87"
-                            // slice = "\xF0\x9F\x87"
-                            // remaining = 1
-                            // ```
-                            //
-                            // We want to compute `endth == 2`:
-                            //
-                            //    2   =      4       -      3      +     1
-                            let endth = substr.len() - slice.len() + remaining;
-                            return Some(&substr[..endth]);
-                        }
-                        // If there are more characters remaining than the number
-                        // of bytes yielded in the invalid UTF-8 byte sequence,
-                        // count `size` bytes and advance the slice to continue
-                        // decoding.
-                        (None, size) => {
-                            slice = &slice[size..];
-                            remaining -= size;
-                        }
-                    }
-                }
-            }
-        }
+        self.inner.get_char_slice(range)
```

Comment:
It feels like the encoded impls should be doing the above (early exit logic etc), however it is pretty irrelevant to the encoding.

Reply:
I agree with your instinct here. It'd also allow us to only pay the branching cost once, for example we branch in `char_len` above in the short circuiting logic.

Happy for this to be addressed in a followup.

I think the short circuiting logic is most useful for UTF-8 case since that is the expensive one to compute.

Reply:
not only in the `char_len`, but also in `len()` too (double cost). So yeah, #future for sure. It feels like `String` in an ideal world would just redirect all calls to `EncodedString`, and if the logic is duplicated in the encoded type, then so be it. This will be much easier/clearer in a follow up PR though as you said :+1:

Reply:
Filed:

- https://github.com/artichoke/artichoke/issues/1722